File: src/main/java/top/craft_hello/tpa/enums/PermissionType.java
package top.craft_hello.tpa.enums;

import org.bukkit.command.CommandSender;

import static java.util.Objects.isNull;

public enum PermissionType {
    DEFAULT, VIP, VIP_PLUS, MVP, MVP_PLUS, MVP_PLUS_PLUS, ADMIN, RELOAD, VERSION, WARP, SET_WARP, DEL_WARP, HOME, SPAWN, SET_SPAWN, DEL_SPAWN, TPA, TP_HERE, TP_ALL, TP_LOGOUT, RTP, DENYS, BACK;

    public static String getPermission(PermissionType permissionType) {
        String permission;
        switch (permissionType){
            case DEFAULT:
                permission = "tpa.default";
                break;
            case VIP:
                permission = "tpa.vip";
                break;
            case VIP_PLUS:
                permission = "tpa.vip+";
                break;
            case MVP:
                permission = "tpa.mvp";
                break;
            case MVP_PLUS:
                permission = "tpa.mvp+";
                break;
            case MVP_PLUS_PLUS:
                permission = "tpa.mvp++";
                break;
            case RELOAD:
                permission = "tpa.reload";
                break;
            case VERSION:
                permission = "tpa.version";
                break;
            case WARP:
                permission = "tpa.warp";
                break;
            case SET_WARP:
                permission = "tpa.setwarp";
                break;
            case DEL_WARP:
                permission = "tpa.delwarp";
                break;
            case HOME:
                permission = "tpa.home";
                break;
            case SPAWN:
                permission = "tpa.spawn";
                break;
            case SET_SPAWN:
                permission = "tpa.setspawn";
                break;
            case DEL_SPAWN:
                permission = "tpa.delspawn";
                break;
            case TPA:
                permission = "tpa.tpa";
                break;
            case TP_HERE:
                permission = "tpa.tphere";
                break;
            case TP_ALL:
                permission = "tpa.tpall";
                break;
            case TP_LOGOUT:
                permission = "tpa.tplogout";
                break;
            case RTP:
                permission = "tpa.rtp";
                break;
            case DENYS:
                permission = "tpa.denys";
                break;
            case BACK:
                permission = "tpa.back";
                break;
            default:
                permission = "tpa.admin";
        }
        return permission;
    }

    public static boolean hasPermission(CommandSender sender, PermissionType... permissionTypes) {
        if (!isNull(sender)) {
            for (PermissionType permissionType : permissionTypes) if (!sender.hasPermission(PermissionType.getPermission(permissionType))) return false;
            return true;
        }
        return false;
    }
}

File: src/main/kotlin/top/craft_hello/tpa/enums/PermissionType.kt
package top.craft_hello.tpa.enums

import org.bukkit.command.CommandSender

enum class PermissionType(val permissionName: String) {
    DEFAULT("tpa.default"),
    VIP("tpa.vip"),
    VIP_PLUS("tpa.vip+"),
    MVP("tpa.mvp"),
    MVP_PLUS("tpa.mvp+"),
    MVP_PLUS_PLUS("tpa.mvp++"),
    ADMIN("tpa.admin"),
    RELOAD("tpa.reload"),
    VERSION("tpa.version"),
    WARP("tpa.warp"),
    SET_WARP("tpa.setwarp"),
    DEL_WARP("tpa.delwarp"),
    HOME("tpa.home"),
    HOMES("tpa.home"),
    SET_HOME("tpa.home"),
    SET_DEFAULT_HOME("tpa.home"),
    DEL_HOME("tpa.home"),
    SPAWN("tpa.spawn"),
    SET_SPAWN("tpa.setspawn"),
    DEL_SPAWN("tpa.delspawn"),
    TPA("tpa.tpa"),
    TP_HERE("tpa.tphere"),
    TP_ALL("tpa.tpall"),
    TP_LOGOUT("tpa.tplogout"),
    RTP("tpa.rtp"),
    DENYS("tpa.denys"),
    BACK("tpa.back"),
    NO_DELAY("tpa.nodelay");

    companion object {
        fun hasPermission(sender: CommandSender, permissionType: PermissionType): Boolean {
            return sender.hasPermission(ADMIN.permissionName) || sender.hasPermission(permissionType.permissionName)
        }

        // 获取玩家所属的等级（用于家数量上限、传送延迟、命令间隔），从高到低逐级判断
        fun getLevel(sender: CommandSender): PermissionType {
            return when {
                hasPermission(sender, ADMIN) -> ADMIN
                hasPermission(sender, MVP_PLUS_PLUS) -> MVP_PLUS_PLUS
                hasPermission(sender, MVP_PLUS) -> MVP_PLUS
                hasPermission(sender, MVP) -> MVP
                hasPermission(sender, VIP_PLUS) -> VIP_PLUS
                hasPermission(sender, VIP) -> VIP
                else -> DEFAULT
            }
        }
    }
}
